### PR TITLE
[Runtimes] Support for aborting Databricks jobs

### PR DESCRIPTION
**Pull request by tomerm-iguazio**
1) Support for aborting Databricks jobs is available.
 When aborting a Databricks job, utilize a Kubernetes hook named `pre_stop` before deleting the runtime's pod. By using this hook, mlrun can execute a script capable of canceling the job within the Databricks environment.
2) User-defined Databricks run names are now supported:
`{"task_parameters": {"databricks_run_name": databricks_run_name}}`
3) An abort test has been added.

[ML-4407](https://jira.iguazeng.com/browse/ML-4407)

### CODE REVIEW

File: mlrun/runtimes/databricks_job/databricks_wrapper.py [hunk 1]
```diff
@@ -40,23 +43,53 @@ def get_task(databricks_run: Run) -> RunTask:
     return databricks_run.tasks[0]
 
 
+def save_credentials(
+    workspace: WorkspaceClient,
+    waiter,
+    host: str,
+    token: str,
+    cluster_id=None,
+    is_finished=False,
```
Comment (gtopper) [inline mlrun/runtimes/databricks_job/databricks_wrapper.py:52]: I don't see any benefit in the default values.
```suggestion
    cluster_id: typing.Optional[str],
    is_finished: bool,
```

**Review by gtopper — CHANGES_REQUESTED** [review]

File: tests/system/runtimes/test_databricks.py [hunk 2]
```diff
@@ -227,3 +276,49 @@ def test_missing_code_run(self):
...
+    def test_abort_task(self):
+        #  in order to clean up any active runs:
+        if self.project.list_runs(state="running"):
+            self.project = mlrun.projects.new_project(self.project_name, overwrite=True)
+        sleep_code = """
+\n
+def handler(**kwargs):
+    import time
+    time.sleep(1000)
+\n
```
[the remark this answers is not in the document]
Comment (tomerm-iguazio) [reply tests/system/runtimes/test_databricks.py:289]: It is just a task that is going to be aborted later, so the sleep length should not be a problem. However, a short sleep period will cause the job to end before we abort it.

File: mlrun/runtimes/databricks_job/databricks_wrapper.py [hunk 3]
```diff
@@ -40,23 +43,53 @@ def get_task(databricks_run: Run) -> RunTask:
...
+    databricks_run = workspace.jobs.get_run(run_id=waiter.run_id)
+    task_run_id = get_task(databricks_run=databricks_run).run_id
+    credentials = {
+        "DATABRICKS_HOST": host,
+        "DATABRICKS_TOKEN": token,
+        "TASK_RUN_ID": task_run_id,
+        "IS_FINISHED": is_finished,
+    }
+    if cluster_id:
+        credentials["DATABRICKS_CLUSTER_ID"] = token
```
[the remark this answers is not in the document]
Comment (tomerm-iguazio) [reply mlrun/runtimes/databricks_job/databricks_wrapper.py:63]: nice catch!

File: tests/system/runtimes/test_databricks.py [hunk 4]
```diff
@@ -227,3 +274,50 @@ def test_missing_code_run(self):
...
+        )
+        function = function_ref.to_function()
+        self._add_databricks_env(function=function)
+        test_run_uid = uuid.uuid4()
+        databricks_run_name = f"databricks_abort_test_{test_run_uid}"
+        params = {"task_parameters": {"databricks_run_name": databricks_run_name}}
+        run_thread = threading.Thread(
+            target=lambda: function.run(project=self.project_name, params=params)
+        )
+        run_thread.start()
```
Comment (gtopper) [inline tests/system/runtimes/test_databricks.py:304]:
```suggestion
        function.run(project=self.project_name, params=params, watch=False)
```

File: tests/system/runtimes/test_databricks.py [hunk 5]
```diff
@@ -227,3 +274,50 @@ def test_missing_code_run(self):
...
+            workspace=workspace, run_name=databricks_run_name
+        )
+        assert run.state.life_cycle_state in (
+            RunLifeCycleState.PENDING,
+            RunLifeCycleState.RUNNING,
+        )
+        self._abort_run()
+        run = workspace.jobs.get_run(run_id=run.run_id)
+        # wait for databricks to update the status.
+        time.sleep(5)
```
Comment (gtopper) [inline tests/system/runtimes/test_databricks.py:318]: Pretty sure you need to sleep before `get_run` for the sleep to have any effect (other than slowing the test down).

File: tests/system/runtimes/test_databricks.py [hunk 6]
```diff
@@ -227,3 +274,50 @@ def test_missing_code_run(self):
                 "Databricks function must be provided with user code"
                 in bad_request_error.value
             )
+
+    def test_abort_task(self):
+        #  in order to clean up any active runs:
```
Comment (gtopper) [inline tests/system/runtimes/test_databricks.py:279]:
```suggestion
        #  in order to clean up any active runs
```

File: tests/system/runtimes/test_databricks.py [hunk 7]
```diff
@@ -227,3 +274,50 @@ def test_missing_code_run(self):
...
+        function = function_ref.to_function()
+        self._add_databricks_env(function=function)
+        test_run_uid = uuid.uuid4()
+        databricks_run_name = f"databricks_abort_test_{test_run_uid}"
+        params = {"task_parameters": {"databricks_run_name": databricks_run_name}}
+        run_thread = threading.Thread(
+            target=lambda: function.run(project=self.project_name, params=params)
+        )
+        run_thread.start()
+        # wait for databricks to run the function.
```
Comment (gtopper) [inline tests/system/runtimes/test_databricks.py:305]:
```suggestion
        # wait for databricks to run the function
```

File: tests/system/runtimes/test_databricks.py [hunk 8]
```diff
@@ -227,3 +274,50 @@ def test_missing_code_run(self):
...
+        function_ref = FunctionReference(
+            kind="databricks",
+            code=sleep_code,
+            image="mlrun/mlrun",
+            name="databricks_abort_test",
+        )
+        function = function_ref.to_function()
+        self._add_databricks_env(function=function)
+        test_run_uid = uuid.uuid4()
+        databricks_run_name = f"databricks_abort_test_{test_run_uid}"
```
Comment (gtopper) [inline tests/system/runtimes/test_databricks.py:299]:
```suggestion
        databricks_run_name = f"databricks_abort_test_{uuid.uuid4()}"
```

File: tests/system/runtimes/test_databricks.py [hunk 9]
```diff
@@ -51,6 +58,46 @@ def %s(**kwargs):
...
+                f"No active runs were found in Databricks with run_name={run_name}"
+            )
+        elif len(runs_by_run_name) > 1:
+            raise MLRunRuntimeError(
+                f"Too many active runs were found in Databricks with run_name={run_name}"
+            )
+        return runs[0]
+
+    def _abort_run(self):
+        print("start aborting")
```
Comment (gtopper) [inline tests/system/runtimes/test_databricks.py:87]: Better use `self._logger`.

File: mlrun/runtimes/databricks_job/databricks_cancel_task.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        print("credentials_path not found. Retrying in 10 seconds...")
+        sleep(10)
+        if not os.path.exists(credentials_path):
+            raise MLRunRuntimeError(
+                "The Databricks credentials path does not exist."
+                " Please manually cancel the job from the Databricks environment."
+            )
+    with open(credentials_path, "r") as yaml_file:
+        loaded_data = yaml.safe_load(yaml_file)
+    #  use for flat structure only.
```
Comment (gtopper) [inline mlrun/runtimes/databricks_job/databricks_cancel_task.py:38]:
```suggestion
    # use for flat yaml only
```

**Review by gtopper — APPROVED** [review]
✅

**Pull request merged**